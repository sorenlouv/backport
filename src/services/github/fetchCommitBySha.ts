import axios from 'axios';
import isEmpty from 'lodash.isempty';
import { BackportOptions } from '../../options/options';
import { HandledError } from '../HandledError';
import { CommitSelected } from './Commit';
import { GithubCommit, GithubSearch } from './GithubApiTypes';
import { handleGithubError } from './handleGithubError';
import { getFormattedCommitMessage } from './commitFormatters';

export async function fetchCommitBySha(
  options: BackportOptions & { sha: string }
): Promise<CommitSelected> {
  const { apiHostname, repoName, repoOwner, sha, accessToken } = options;
  try {
    const res = await axios.get<GithubSearch<GithubCommit>>(
      `https://${apiHostname}/search/commits?q=hash:${sha}%20repo:${repoOwner}/${repoName}&per_page=1&access_token=${accessToken}`,
      {
        headers: {
          Accept: 'application/vnd.github.cloak-preview'
        }
      }
    );

    if (isEmpty(res.data.items)) {
      throw new HandledError(`No commit found on master with sha "${sha}"`);
    }

    const commitRes = res.data.items[0];
    const fullSha = commitRes.sha;

    const message = getFormattedCommitMessage({
      message: commitRes.commit.message,
      sha: fullSha
    });

    return {
      message,
      sha: fullSha
    };
  } catch (e) {
    throw handleGithubError(e);
  }
}
