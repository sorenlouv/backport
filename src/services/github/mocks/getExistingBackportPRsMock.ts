export const getExistingBackportPRsMock = (repoName: string) => ({
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        repository: {
                          name: repoName,
                          owner: {
                            login: 'elastic',
                          },
                        },
                        number: 80,
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                source: {
                                  __typename: 'PullRequest',
                                  state: 'MERGED',
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          commit: {
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
          ],
        },
      },
    },
  },
});
