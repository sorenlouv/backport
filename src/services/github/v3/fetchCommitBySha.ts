import axios, { AxiosResponse } from 'axios';
import isEmpty from 'lodash.isempty';
import { BackportOptions } from '../../../options/options';
import { HandledError } from '../../HandledError';
import { CommitSelected } from '../Commit';
import { handleGithubV3Error } from './handleGithubV3Error';
import { getFormattedCommitMessage } from '../commitFormatters';

export interface GithubCommit {
  commit: {
    message: string;
  };
  sha: string;
}

interface GithubSearch<T> {
  items: T[];
}

export async function fetchCommitBySha(
  options: BackportOptions & { sha: string }
): Promise<CommitSelected> {
  const {
    githubApiBaseUrlV3,
    repoName,
    repoOwner,
    sha,
    accessToken,
    username,
  } = options;

  let res: AxiosResponse<GithubSearch<GithubCommit>>;
  try {
    res = await axios.get<GithubSearch<GithubCommit>>(
      `${githubApiBaseUrlV3}/search/commits?q=hash:${sha}%20repo:${repoOwner}/${repoName}&per_page=1`,
      {
        auth: {
          username: username,
          password: accessToken,
        },
        headers: {
          Accept: 'application/vnd.github.cloak-preview',
        },
      }
    );
  } catch (e) {
    throw handleGithubV3Error(e);
  }

  // TODO: it should be possible to backport from other branches than master
  if (isEmpty(res.data.items)) {
    throw new HandledError(`No commit found on master with sha "${sha}"`);
  }

  const commitRes = res.data.items[0];
  const fullSha = commitRes.sha;

  const formattedMessage = getFormattedCommitMessage({
    message: commitRes.commit.message,
    sha: fullSha,
  });

  return {
    branch: 'master',
    formattedMessage,
    sha: fullSha,
  };
}
