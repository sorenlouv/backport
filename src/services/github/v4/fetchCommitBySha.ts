import chalk from 'chalk';
import ora from 'ora';
import { BackportOptions } from '../../../options/options';
import { Commit } from '../../../types/Commit';
import { HandledError } from '../../HandledError';
import {
  getFormattedCommitMessage,
  getShortSha,
  getPullNumberFromMessage,
} from '../commitFormatters';
import { apiRequestV4 } from './apiRequestV4';
import {
  PullRequestNode,
  pullRequestFragment,
  pullRequestFragmentName,
  getExistingTargetPullRequests,
  getPullRequestLabels,
} from './getExistingTargetPullRequests';
import { getTargetBranchesFromLabels } from './getTargetBranchesFromLabels';

export async function fetchCommitBySha(
  options: BackportOptions & { sha: string }
): Promise<Commit> {
  const { accessToken, githubApiBaseUrlV4, repoName, repoOwner } = options;

  const query = /* GraphQL */ `
  query CommitsBySha($repoOwner: String!, $repoName: String!, $oid: String!) {
    repository(owner: $repoOwner, name: $repoName) {
      object(expression: $oid) {
        ... on Commit {
          message
          oid
          associatedPullRequests(first: 1) {
            edges {
              node {
                ...${pullRequestFragmentName}
              }
            }
          }
        }
      }
    }
  }

    ${pullRequestFragment}
  `;

  const spinner = ora(`Loading commit "${getShortSha(options.sha)}"`).start();

  let res: DataResponse;
  try {
    res = await apiRequestV4<DataResponse>({
      githubApiBaseUrlV4,
      accessToken,
      query,
      variables: {
        repoOwner,
        repoName,
        oid: options.sha,
      },
    });
    spinner.stop();
  } catch (e) {
    spinner.fail();
    throw e;
  }

  if (!res.repository.object) {
    throw new HandledError(
      `No commit found on branch "${options.sourceBranch}" with sha "${options.sha}"`
    );
  }

  const sha = res.repository.object.oid;
  const commitMessage = res.repository.object.message;
  const pullRequestNode =
    res.repository.object.associatedPullRequests.edges?.[0]?.node;

  const pullNumber =
    pullRequestNode?.number || getPullNumberFromMessage(commitMessage);

  const sourceBranch = pullRequestNode?.baseRefName || options.sourceBranch;

  const formattedMessage = getFormattedCommitMessage({
    message: commitMessage,
    pullNumber,
    sha,
  });

  spinner.stopAndPersist({
    symbol: chalk.green('?'),
    text: `${chalk.bold('Select commit')} ${chalk.cyan(formattedMessage)}`,
  });

  const existingTargetPullRequests = getExistingTargetPullRequests(
    commitMessage,
    pullRequestNode
  );

  const targetBranchesFromLabels = getTargetBranchesFromLabels({
    existingTargetPullRequests,
    branchLabelMapping: options.branchLabelMapping,
    labels: getPullRequestLabels(pullRequestNode),
  });

  return {
    sourceBranch,
    targetBranchesFromLabels,
    sha,
    formattedMessage,
    originalMessage: commitMessage,
    pullNumber,
    existingTargetPullRequests,
  };
}

interface DataResponse {
  repository: {
    object: {
      message: string;
      oid: string;
      associatedPullRequests: {
        edges:
          | {
              node: PullRequestNode;
            }[]
          | null;
      };
    } | null;
  };
}
