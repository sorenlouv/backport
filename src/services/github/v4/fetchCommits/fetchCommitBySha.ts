import { ValidConfigOptions } from '../../../../options/options';
import { HandledError } from '../../../HandledError';
import {
  Commit,
  SourceCommitWithTargetPullRequest,
  sourceCommitWithTargetPullRequestFragment,
  parseSourceCommit,
} from '../../../sourceCommit/parseSourceCommit';
import { apiRequestV4 } from '../apiRequestV4';

export async function fetchCommitBySha(options: {
  accessToken: string;
  branchLabelMapping?: ValidConfigOptions['branchLabelMapping'];
  githubApiBaseUrlV4?: string;
  historicalBranchLabelMappings: ValidConfigOptions['historicalBranchLabelMappings'];
  repoName: string;
  repoOwner: string;
  sha: string;
  sourceBranch: string;
}): Promise<Commit> {
  const {
    accessToken,
    githubApiBaseUrlV4 = 'https://api.github.com/graphql',
    repoName,
    repoOwner,
    sha,
    sourceBranch,
  } = options;

  const query = /* GraphQL */ `
    query CommitsBySha($repoOwner: String!, $repoName: String!, $oid: String!) {
      repository(owner: $repoOwner, name: $repoName) {
        object(expression: $oid) {
          ...SourceCommitWithTargetPullRequest
        }
      }
    }

    ${sourceCommitWithTargetPullRequestFragment.source}
  `;

  const res = await apiRequestV4<CommitsByShaResponse>({
    githubApiBaseUrlV4,
    accessToken,
    query,
    variables: {
      repoOwner,
      repoName,
      oid: sha,
    },
  });

  const sourceCommit = res.repository.object;
  if (!sourceCommit) {
    throw new HandledError(
      `No commit found on branch "${sourceBranch}" with sha "${sha}"`
    );
  }

  return parseSourceCommit({ options, sourceCommit });
}

interface CommitsByShaResponse {
  repository: {
    object: SourceCommitWithTargetPullRequest | null;
  };
}
