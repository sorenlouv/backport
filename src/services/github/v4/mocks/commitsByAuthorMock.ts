import { DataResponse } from '../fetchCommitsByAuthor';

export const commitsWithPullRequestsMock: DataResponse = {
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                oid: '2e63475c483f7844b0f2833bc57fdee32095bacb',
                message: 'Add ðŸ‘»',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                message: 'Add witch (#85)',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        baseRefName: 'master',
                        labels: {
                          nodes: [{ name: 'my-label-b' }],
                        },
                        mergeCommit: {
                          oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                          message: 'Add witch (#85)',
                        },
                        repository: {
                          name: 'kibana',
                          owner: {
                            login: 'elastic',
                          },
                        },
                        number: 85,
                        timelineItems: {
                          edges: [],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        labels: {
                          nodes: [{ name: 'my-label-a' }],
                        },
                        mergeCommit: {
                          oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                          message:
                            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                        },
                        repository: {
                          name: 'kibana',
                          owner: {
                            login: 'elastic',
                          },
                        },
                        baseRefName: 'master',
                        number: 80,
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                source: {
                                  __typename: 'PullRequest',
                                  title: 'some title',
                                  state: 'MERGED',
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          commit: {
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                oid: '3827bbbaf39914eda4f02f6940189844375fd097',
                message: 'Add backport config',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                oid: '5ea0da550ac191029459289d67f99ad7d310812b',
                message: 'Initial commit',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
          ],
        },
      },
    },
  },
};
