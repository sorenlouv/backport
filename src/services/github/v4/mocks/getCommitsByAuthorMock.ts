import { CommitByAuthorResponse } from '../fetchCommitsByAuthor';

export const getCommitsByAuthorMock = (
  repoName: string
): CommitByAuthorResponse => ({
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        baseRefName: 'master',
                        labels: {
                          nodes: [{ name: 'my-label-a' }],
                        },
                        mergeCommit: {
                          oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                          message:
                            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                        },
                        repository: {
                          name: repoName,
                          owner: {
                            login: 'elastic',
                          },
                        },
                        number: 80,
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                source: {
                                  __typename: 'PullRequest',
                                  number: 99,
                                  title: 'some title',
                                  state: 'MERGED',
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          commit: {
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
          ],
        },
      },
    },
  },
});
