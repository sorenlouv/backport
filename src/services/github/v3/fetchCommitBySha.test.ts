import axios from 'axios';
import { BackportOptions } from '../../../options/options';
import { CommitSelected } from '../../../types/Commit';
import { fetchCommitBySha } from './fetchCommitBySha';
import { commitByShaMock } from './mocks/commitByShaMock';

describe('fetchCommitBySha', () => {
  it('should return single commit with pull request', async () => {
    const commitSha = 'sha123456789';
    const options = {
      repoOwner: 'elastic',
      repoName: 'kibana',
      accessToken: 'myAccessToken',
      username: 'sqren',
      author: 'sqren',
      githubApiBaseUrlV3: 'https://api.github.com',
    } as BackportOptions;

    const axiosSpy = jest
      .spyOn(axios, 'request')

      // mock commits
      .mockResolvedValueOnce({
        data: { items: [{ commit: { message: 'myMessage' }, sha: commitSha }] },
      });

    await expect(
      await fetchCommitBySha({ ...options, sha: commitSha })
    ).toEqual({
      sourceBranch: 'master',
      formattedMessage: 'myMessage (sha12345)',
      pullNumber: undefined,
      sha: 'sha123456789',
      targetBranchesFromLabels: [],
    });

    expect(axiosSpy.mock.calls).toMatchSnapshot();
  });

  it('should return a single commit without PR', async () => {
    const axiosSpy = jest
      .spyOn(axios, 'request')
      .mockResolvedValueOnce({ data: { items: [commitByShaMock] } });

    const commit = await fetchCommitBySha({
      username: 'sqren',
      accessToken: 'myAccessToken',
      repoOwner: 'elastic',
      repoName: 'kibana',
      sha: 'myCommitSha',
      githubApiBaseUrlV3: 'https://api.github.com',
    } as BackportOptions & { sha: string });

    const expectedCommit: CommitSelected = {
      sourceBranch: 'master',
      formattedMessage:
        '[Chrome] Bootstrap Angular into document.body (myCommit)',
      sha: 'myCommitSha',
      pullNumber: undefined,
      targetBranchesFromLabels: [],
    };

    expect(commit).toEqual(expectedCommit);
    expect(axiosSpy).toHaveBeenCalledWith({
      method: 'get',
      url:
        'https://api.github.com/search/commits?q=hash:myCommitSha%20repo:elastic/kibana&per_page=1',
      headers: { Accept: 'application/vnd.github.cloak-preview' },
      auth: { password: 'myAccessToken', username: 'sqren' },
    });
  });

  it('should throw error if sha does not exist', async () => {
    jest.spyOn(axios, 'request').mockResolvedValueOnce({ data: { items: [] } });

    await expect(
      fetchCommitBySha({
        repoOwner: 'elastic',
        repoName: 'kibana',
        sha: 'myCommitSha',
        githubApiBaseUrlV3: 'https://api.github.com',
      } as BackportOptions & { sha: string })
    ).rejects.toThrowError('No commit found on master with sha "myCommitSha"');
  });
});
