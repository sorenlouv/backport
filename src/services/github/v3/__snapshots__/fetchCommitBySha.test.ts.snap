// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCommitBySha should return single commit with pull request 1`] = `
Array [
  Array [
    "https://api.github.com/search/commits?q=hash:sha123456789%20repo:elastic/kibana&per_page=1",
    Object {
      "auth": Object {
        "password": "myAccessToken",
        "username": "sqren",
      },
      "headers": Object {
        "Accept": "application/vnd.github.cloak-preview",
      },
    },
  ],
]
`;
